Camera lookAt

bagnell opened the pull request:
Deprecated the `eye`, `target`, and `up` parameters to `Camera.lookAt`.

`Camera.lookAt` now sets the camera position and orientation using a target, offset and transformation matrix. The target must be given in world coordinates. The offset is a cartesian offset from the target in a reference frame centered by the target. By default, the reference frame is the local east-north-up frame centered at the target. The transformation matrix can be used to specify a different reference frame.

pjcozzi commented on CHANGES.md:
```diff
@@ -12,6 +12,7 @@ Change Log
   * `Camera.tilt` was deprecated in Cesium 1.6. It will be removed in Cesium 1.7. Use `Camera.pitch`.
   * `Camera.heading` and `Camera.tilt` were deprecated in Cesium 1.6. They will become read-only in Cesium 1.7. Use `Camera.setView`.
   * `Camera.setPositionCartographic` was deprecated in Cesium 1.6. It will be removed in Cesium 1.7. Use `Camera.setView`.
+   * The `eye`, `target`, and `up` parameters to `Camera.lookA`t were deprecated in Cesium 1.6. It will be removed in Cesium 1.8. Use the `target`, `offset`, and `transform` parameters.
```
Submit an issue.  In general, just submit them when you open the pull request.  If we decide against them, we'll just close the issue.

pjcozzi commented on Apps/Sandcastle/gallery/Camera.html:
```diff
@@ -108,14 +108,15 @@
+        offset : new Cesium.Cartesian3(-120000.0, -120000.0, 120000.0),
+        transform : transform
+    });
+    camera.setTransform(transform);
```
It's not clear to me why we need to call `setTransform` here but not in the `3D Models.html` above.  Perhaps one is wrong?  In general, will users be able to figure out what it means to pass a `transform` to `lookAt` vs. call `setTransform`?

bagnell replied:
`3D Models.html` sets it too. Its just before the call to `lookAt`. In both cases, the call to `lookAt` doesn't need the `transform` because its already what will be done by default.

pjcozzi replied:
Ah, it sets the property.  Err, that is confusing.  We have a property, a function, and a function parameter.  Are we sure about this?

bagnell replied:
Yeah, I agree. They all do things differently though. I'm open to suggestions. Perhaps better names?

pjcozzi replied:
Err, you'll have to explain what each of these do.  The reference doc is too in the mud.  I suspect that we need some redesign across the `camera` as some things are no longer needed and there is probably more state than necessary.

pjcozzi commented on Source/Scene/Camera.js:
```diff
@@ -1527,59 +1527,164 @@ define([
+     *
+     * @param {Object} options An object with the following properties:
+     * @param {Cartesian3} options.target The target position in world coordinates.
+     * @param {Cartesian3} options.offset The offset from the target in a reference frame centered at the target.
```
Is this the frame defined by `transform`?  `"centered at the target"` is not enough info.

pjcozzi commented on Source/Scene/Camera.js:
```diff
@@ -1527,59 +1527,164 @@ define([
+     * @param {Object} options An object with the following properties:
+     * @param {Cartesian3} options.target The target position in world coordinates.
+     * @param {Cartesian3} options.offset The offset from the target in a reference frame centered at the target.
+     * @param {Matrix4} [options.transform] The transformation matrix defining the reference frame centered at the target. The default is a local east-north-up reference frame.
```
When `transform` is defined, is `target` ignored?  It seems like this is ambiguous since the translation is basically provided in separate places.  It almost feels like this should be a 3x3 matrix or we should allow `(target && offset)` or `(offset && transform)` perhaps even using two separate functions.

bagnell replied:
That's a good point. I'll create two separate functions.

pjcozzi commented:
Does this require any updates to the [camera tutorial](https://cesiumjs.org/2013/02/13/Cesium-Camera-Tutorial/)?

bagnell commented:
>Does this require any updates to the camera tutorial?

Nothing needs to be updated, but I'm going to add a section for this since I think it will be more widely used now.

bagnell commented:
@pjcozzi This is ready for another review. 

I plan to deprecate the `transform` property and the `setTransform` function. I'd make the `transform` a read-only property and remove the function (though I'd keep a private version because its useful internally). I want to do it in a separate PR. I'm not sure how much these changes would affect.

bagnell commented:
@pjcozzi This is ready for another review. I'll summarize the changes since last time you looked at it:

* Added `HeadingPitchRange` which has heading and pitch angle properties in radians and a range in meters.
* `Camera.lookAt` now takes a target and an offset. The offset can be be either a `Cartesian3` or a `HeadingPitchRange`. The offset is in the east-north-up reference frame centered at target.
* Added `Camera.lookAtTransform` which is the same as `lookAt`, but takes a 4x4 matrix instead of a target and east-north-up frame.
* Deprecated the `Camera.setTransform` function and the `Camera.transform` property.

pjcozzi commented:
Beautiful.  Let the forum know this is coming in 1.6.
